README.md
# VoiceAssistant

<p align="center"><img src="https://raw.githubusercontent.com/Mertalmali4/VoiceAssistant/master/icon/AI2X.png" width="200" height="240"></p>

<div align="center">

<a>This Voice Assistant written in Python.Voice Assistant includes Face Recognition for Authentication, Hand Controlled Mouse and Hand Controlled Volume Up&Down now</a>

<a>These features will be developed and new features will continue to be added</a>

<a>Don't forget to let us know any problems you see. have fun!</a>
















Changes.txt
C:\ProgramData\Anaconda3\lib\site-packages\comtypes\__init__.py,
152. satıra try except ekledim




ip.py
'''
#-*-coding:utf-8-*-

#Sorgu işleminde kullanılacak modüller:
#(eğer modüller yüklü değil ise komut sistemine "pip install requests" komutunu kullanarak yükleyebilirsiniz)
import requests, json

#Sorgu sonucu listelenecek başlıklar:
print ("IP","Ülke","İl","İlçe","Enlem","Boylam")

#Sorgulanacak ip lere ait liste:
ip = ["176.220.33.233"]
APIKEY = "81ac0d765ed9122937ae64cbae0c1e80"
#ip listesinin teker teker web servis üzerinden sorgulanması:
for x in ip:
  #Bu kısımda yer alan API_KEY'i https://ipstack.com/ adresine üye olarak temin edebilirsiniz.
  serviceURL = "http://api.ipapi.com/"+x+"?access_key="+APIKEY+"&output=json"  
  r = requests.get(serviceURL)
  y = json.loads(r.text)
  print(y)
  #print(y["ip"],y["country_name"],y["region_name"],y["city"],y["latitude"],y["longitude"])  



  '''




import phonenumbers
from phonenumbers import carrier,timezone,geocoder

number = input("Telefon Numarasını  Gir: ")

if number.startswith("+"):
    print("Numara Doğru!")
    Numara = phonenumbers.parse(number)
    zaman = timezone.time_zones_for_number(Numara)
    sim_adi = carrier.name_for_number(Numara, "tr")
    bolge = geocoder.description_for_number(Numara, "tr")

    print("Saat Dilimi :", zaman)
    print("Sim adı:", sim_adi)
    print("Yaşadığı Ülke(bölge) :", bolge)

else:
    print("+ülke kodunu kullanarak giriniz.")

training.py
import speech_recognition as sr
import time
import sys
from picovoice import Picovoice
from pvrecorder import PvRecorder
import time
import sqlite3
from playsound import playsound
import wikipedia
import speech_recognition as sr
import pyttsx3
from datetime import datetime
import webbrowser
import wolframalpha
import pvporcupine
from threading import Thread
from retinaface import RetinaFace
import face_recognition
import cv2
import numpy as np





'''
def callback(recognizer, audio):                          # this is called from the background thread
    try:
        text = recognizer.recognize_google(audio, language = 'en-IN', show_all = True )
        print("I thinks you said '" + recognizer.recognize_google(audio) + "'")

        #print("You said " + recognizer.recognize_google(audio))  # received audio data, now need to recognize it
    except sr.UnknownValueError:
        print("Oops! Didn't catch that")
    except LookupError:
        print("Oops! Didn't catch that")
r = sr.Recognizer()
r.listen_in_background(sr.Microphone(), callback)

while True: time.sleep(0.1)  

'''

'''
def standby():
    r=sr.Recognizer()
    with sr.Microphone() as source:
        print("listening...")
        r.pause_threshold=0.6
        try:

            audio=r.listen(source,timeout=2,phrase_time_limit=7)

        except sr.WaitTimeoutError:
            return "none"
    try:
        print("recognizing...")
        query=r.recognize_google(audio,language="en-US")
        print(f":{query}")
    except sr.WaitTimeoutError:
        return "none"
    except Exception as e:
        return "none"
    
    

    query =query.lower()
    return query

while True:
    permission=standby()
    if("hello"in permission):
        print("okey")
    elif("bye"in permission):
        sys.exit()

'''


import pvporcupine
'''
# AccessKey obtained from Picovoice Console (https://console.picovoice.ai/)
access_key = "8fPjSJyCB9Lb/4UE8Zza3F0/XaC5cE31KbBAjAiEjiTUsbd8UX1OKw=="

handle = pvporcupine.create(access_key=access_key, keywords=['jarvis', 'bumblebee'])
'''


'''
def get_next_audio_frame():
  
  pass

while True:
  
  audio_frame = get_next_audio_frame()
  keyword_index = porcupine.process(audio_frame)
  if keyword_index >= 0:
      print("detected")
      pass
      '''
'''

def wake_word_callback():
    pass
context_path = ...  # path to Rhino context file (.RHN)
def inference_callback(inference):
    print(inference.is_understood)
    if inference.is_understood:
        print(inference.intent)
        for k, v in inference.slots.items():
            print(f"{k} : {v}")

pv = Picovoice(
    access_key={"8fPjSJyCB9Lb/4UE8Zza3F0/XaC5cE31KbBAjAiEjiTUsbd8UX1OKw=="},
    keyword_path="jarvis_en_windows_v2_1_0 -.ppn",
    wake_word_callback=wake_word_callback,
    context_path=context_path(),
    inference_callback=inference_callback)

porcupine = pvporcupine.create(
  access_key="8fPjSJyCB9Lb/4UE8Zza3F0/XaC5cE31KbBAjAiEjiTUsbd8UX1OKw==",
  keyword_paths=['jarvis_en_windows_v2_1_0 -.ppn']
)

'''
'''
import pvporcupine


for keyword in pvporcupine.KEYWORDS:
    print(keyword)








from pvrecorder import PvRecorder

porcupine = pvporcupine.create(
  access_key="8fPjSJyCB9Lb/4UE8Zza3F0/XaC5cE31KbBAjAiEjiTUsbd8UX1OKw==",
  keyword_paths=['jarvis_en_windows_v2_1_0 -.ppn'])
recoder = PvRecorder(device_index=-1, frame_length=porcupine.frame_length)





try:
    recoder.start()
    while True:
        keyword_index = porcupine.process(recoder.read())
        if keyword_index >= 0:
            print("detected")
except KeyboardInterrupt:
    recoder.stop()
finally:
    porcupine.delete()
    recoder.delete()

'''

'''
#phone number location
import phonenumbers
from phonenumbers import carrier,timezone,geocoder

number = input("Telefon Numarasını  Gir: ")

if number.startswith("+"):
    print("Numara Doğru!")
    Numara = phonenumbers.parse(number)
    zaman = timezone.time_zones_for_number(Numara)
    sim_adi = carrier.name_for_number(Numara, "tr")
    bolge = geocoder.description_for_number(Numara, "tr")

    print("Saat Dilimi :", zaman)
    print("Sim adı:", sim_adi)
    print("Yaşadığı Ülke(bölge) :", bolge)

else:
    print("+ülke kodunu kullanarak giriniz.")

'''





video_capture = cv2.VideoCapture(0)
    

    

    # Load a second sample picture and learn how to recognize it.
mert_image = cv2.imread("dataset/mert.jpg")
mert_face_encoding = face_recognition.face_encodings(mert_image)[0]

    # Create arrays of known face encodings and their names
known_face_encodings = [
            
mert_face_encoding
]
known_face_names = [
    
    "Mert"
]

face_locations = []
face_encodings = []
face_names = []
process_this_frame = True
while True:
    
    # Grab a single frame of video
    ret, frame = video_capture.read()
    # Only process every other frame of video to save time
    if process_this_frame:
        # Resize frame of video to 1/4 size for faster face recognition processing
        small_frame = cv2.resize(frame, (0, 0), fx=0.25, fy=0.25)
        # Convert the image from BGR color (which OpenCV uses) to RGB color (which face_recognition uses)
        rgb_small_frame = small_frame[:, :, ::-1]
        
        # Find all the faces and face encodings in the current frame of video
        face_locations = face_recognition.face_locations(rgb_small_frame)
        face_encodings = face_recognition.face_encodings(rgb_small_frame, face_locations)
        face_names = []
        for face_encoding in face_encodings:
            # See if the face is a match for the known face(s)
            matches = face_recognition.compare_faces(known_face_encodings, face_encoding,tolerance=0.5)
            name = "Unknown"
            # # If a match was found in known_face_encodings, just use the first one.
            # if True in matches:
            #     first_match_index = matches.index(True)
            #     name = known_face_names[first_match_index]
            # Or instead, use the known face with the smallest distance to the new face
            face_distances = face_recognition.face_distance(known_face_encodings, face_encoding)
            best_match_index = np.argmin(face_distances)
            if matches[best_match_index]:
                name = known_face_names[best_match_index]
            face_names.append(name)


    process_this_frame = not process_this_frame
    # Display the results
    for (top, right, bottom, left), name in zip(face_locations, face_names):
        # Scale back up face locations since the frame we detected in was scaled to 1/4 size
        top *= 4
        right *= 4
        bottom *= 4
        left *= 4
        # Draw a box around the face
        cv2.rectangle(frame, (left, top), (right, bottom), (0, 0, 255), 2)
        # Draw a label with a name below the face
        cv2.rectangle(frame, (left, bottom - 35), (right, bottom), (0, 0, 255), cv2.FILLED)
        font = cv2.FONT_HERSHEY_DUPLEX
        cv2.putText(frame, name, (left + 6, bottom - 6), font, 1.0, (255, 255, 255), 1)
    # Display the resulting image
    cv2.imshow("Video", frame)
    # Hit 'q' on the keyboard to quit!
    if("Mert" in face_names):
        print("başarılı")
    
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break
# Release handle to the webcam



video_capture.release()
cv2.destroyAllWindows()
        
        
        

#a=face_recognition()








hand_pose.py
import cv2
import mediapipe as mp
import mouse


mp_drawing = mp.solutions.drawing_utils
mp_drawing_styles = mp.solutions.drawing_styles
mp_hands = mp.solutions.hands

index_finger_mcp=0
index_finger_tip = 0
middle_finger_tip=0
thumb_tip = 0
distance = 0
y_axis=0
sensitivity=3
j=0


cap = cv2.VideoCapture(0)
with mp_hands.Hands(
    max_num_hands=1,
    model_complexity=0,
    min_detection_confidence=0.55,
    min_tracking_confidence=0.55) as hands:
  
    

    
  while cap.isOpened():
    success, image = cap.read()
    if not success:
      print("Ignoring empty camera frame.")
      # If loading a video, use 'break' instead of 'continue'.
      continue

    # To improve performance, optionally mark the image as not writeable to
    # pass by reference.
    image.flags.writeable = False
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    results = hands.process(image)

    # Draw the hand annotations on the image.
    image.flags.writeable = True
    image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
    if results.multi_hand_landmarks:
      for hand_landmarks in results.multi_hand_landmarks:
        mp_drawing.draw_landmarks(
            image,
            hand_landmarks,
            mp_hands.HAND_CONNECTIONS,
            mp_drawing_styles.get_default_hand_landmarks_style(),
            mp_drawing_styles.get_default_hand_connections_style())
    # Flip the image horizontally for a selfie-view display.
    
    if not results.multi_hand_landmarks:
      continue
    image_height, image_width, _ = image.shape
    annotated_image = image.copy()
    for hand_landmarks in results.multi_hand_landmarks:
      

      index_finger_tip = hand_landmarks.landmark[mp_hands.HandLandmark.INDEX_FINGER_TIP]
      thumb_tip = hand_landmarks.landmark[mp_hands.HandLandmark.THUMB_TIP]
      index_finger_mcp= hand_landmarks.landmark[mp_hands.HandLandmark.MIDDLE_FINGER_MCP]
      middle_finger_tip= hand_landmarks.landmark[mp_hands.HandLandmark.MIDDLE_FINGER_TIP]
      
      rescaled_index_finger_tip = (int(index_finger_tip.x * image_width),int(index_finger_tip.y * image_height))
      rescaled_thumb_tip = (int(thumb_tip.x * image_width), int(thumb_tip.y * image_height))
      rescaled_middle_finger_tip = (int(middle_finger_tip.x * image_width), int(middle_finger_tip.y * image_height))

      distance = ((rescaled_index_finger_tip[0] - rescaled_thumb_tip[0])**2 + (rescaled_index_finger_tip[1] - rescaled_thumb_tip[1])**2)**0.5
      distanceclick= ((rescaled_index_finger_tip[0] - rescaled_middle_finger_tip[0])**2 + (rescaled_index_finger_tip[1] - rescaled_middle_finger_tip[1])**2)**0.5
      
      #x =1980-(int(index_finger_tip.x * image_width)*4)
      
      y=int((index_finger_mcp.y*1700)-280)
      
      x =2000-(int(index_finger_mcp.x * image_width)*5)
      
      

      '''
      if((y_axis-y)<=0):
        if((y_axis-y)==0):
          a=mouse.get_position()
          a=a[1]
        a=mouse.get_position()
        a=a[1]+((y-y_axis)*sensitivity)
        y_axis=y
      
      else:
        a=mouse.get_position()
        a=a[1]-((y+y_axis)*sensitivity)
        y_axis=y

      '''
      mouse.move(x,y,absolute=True,duration=0.02)



      z = ((index_finger_mcp.z)*-100)
      print(distance,z)

      
      if(z>=5):
        if(distanceclick<26):
          mouse.click("left")
            
        if(distance<27):
          mouse.hold("left")
          break
        if(distance>40):
          mouse.release("left")
      if(z>=3):
        
        if(distanceclick<26):
          mouse.click("left")
            
        if(distance<27):
          mouse.hold("left")
          break
        if(distance>40):
          mouse.release("left")
      if(z>=1):
        
        if(distanceclick<22):
          mouse.click("left")
            
        if(distance<24):
          mouse.hold("left")
          break
        if(distance>35):
          mouse.release("left")
      if(z<=1):
        
        if(distanceclick<22):
          mouse.click("left")
            
        if(distance<24):
          mouse.hold("left")
          break
        if(distance>35):
          mouse.release("left")
       
      



      
      
      #mouse.move(x,y,absolute=True,duration=0.07)
      
      
          
    
      

      


      mp_drawing.draw_landmarks(
          annotated_image,
          hand_landmarks,
          mp_hands.HAND_CONNECTIONS,
          mp_drawing_styles.get_default_hand_landmarks_style(),
          mp_drawing_styles.get_default_hand_connections_style())
    
    # Draw hand world landmarks.
    if not results.multi_hand_world_landmarks:
      continue







    cv2.imshow('MediaPipe Hands', cv2.flip(image, 1))
    if cv2.waitKey(5) & 0xFF == 27:
      break
cap.release()







head_pose.py
import cv2
import mediapipe as mp
import mouse
import numpy as np

mp_drawing = mp.solutions.drawing_utils
mp_face_mesh = mp.solutions.face_mesh
face_mesh = mp_face_mesh.FaceMesh(min_detection_confidence=0.5,min_tracking_confidence=0.5)

drawing_spec = mp_drawing.DrawingSpec(thickness = 1,circle_radius = 1)

cap = cv2.VideoCapture(0)


while cap.isOpened():
    sucess, image = cap.read()


    image = cv2.cvtColor(cv2.flip(image,1),cv2.COLOR_BGR2RGB)

    image.flags.writeable = False

    results = face_mesh.process(image)

    image.flags.writeable = True

    image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)

    img_h,img_w,img_c = image.shape

    face_3d = []
    face_2d = []




    if results.multi_face_landmarks:
        for face_landmarks in results.multi_face_landmarks:
            for idx, lm in enumerate(face_landmarks.landmark):
                if idx == 33 or idx==263 or idx==1 or idx==61 or idx==291 or idx==199:
                    if idx == 1:
                        nose_2d = (lm.x * img_w,lm.y * img_h)
                        nose_3d = (lm.x * img_w,lm.y * img_h,lm.z * 3000)

                    x, y = int(lm.x * img_w),int(lm.y * img_h)


                    face_2d.append([x,y])

                    face_3d.append([x,y,lm.z])


            face_2d = np.array(face_2d,dtype=np.float64)
                
            face_3d = np.array(face_3d,dtype=np.float64)        


            focal_length = 1 * img_w

            cam_matrix = np.array([ [focal_length,0,img_h / 2],
                                   [0,focal_length,img_w /2],
                                   [0,0,1]])
            
            dist_matrix = np.zeros((4,1),dtype=np.float64)
            sucess, rot_vec, trans_vec = cv2.solvePnP(face_3d,face_2d,cam_matrix,dist_matrix)


            rmat, jac = cv2.Rodrigues(rot_vec)

            angles, mtxR, mtxQ,Qx,Qy,Qz=cv2.RQDecomp3x3(rmat)

            x=angles[0] * 360
            y=angles[1] * 360
            z=angles[2] * 360

            nose_3d_projection, jacobian = cv2.projectPoints(nose_3d,rot_vec,trans_vec,cam_matrix,dist_matrix)

            p1 = (int(nose_2d[0]),int(nose_2d[1]))
            p2 = (int(nose_2d[0] + y*10),int(nose_2d[1]-x*10))


            cv2.line(image,p1,p2,(255,0,0),3)



            x_mouse=int(np.round(x,2))*140
            y_mouse =int(np.round(x,2))*50
            print(int(np.round(x,2)))
            mouse.move(x_mouse,y_mouse,absolute=True,duration=0.05)

            cv2.putText(image,"x: "+str(np.round(x,2)),(500,50),cv2.FONT_HERSHEY_SIMPLEX,1,(0,0,255),1)
            cv2.putText(image,"y: "+str(np.round(y,2)),(500,100),cv2.FONT_HERSHEY_SIMPLEX,1,(0,0,255),1)
            cv2.putText(image,"z: "+str(np.round(z,2)),(500,150),cv2.FONT_HERSHEY_SIMPLEX,1,(0,0,255),1)

            mp_drawing.draw_landmarks(
                image=image,
                landmark_list=face_landmarks,
                connections=mp_face_mesh.FACEMESH_TESSELATION,
                landmark_drawing_spec=drawing_spec,
                connection_drawing_spec=drawing_spec)
            
        cv2.imshow('Head Pose Estimation',image)


        if cv2.waitKey(5) & 0xFF ==27:
            break


cap.release()

voice_assistant.py
from pvrecorder import PvRecorder
import time
import sqlite3
from playsound import playsound
import wikipedia
import speech_recognition as sr
import pyttsx3
from datetime import datetime
import webbrowser
import wolframalpha
import pvporcupine
from threading import Thread
from retinaface import RetinaFace
import face_recognition
import cv2
import numpy as np
import sys
import random
import pyjokes
from selenium import webdriver
from selenium.webdriver.common.by import By



class Voice_Assistant():


    
    
    


    def __init__(self):
        super().__init__()
        self.i = 0
        self.rate=185
        self.youtube()
        
        

        
    

    def speak(self,say):
        
        self.engine = pyttsx3.init()
        self.engine.setProperty("rate", self.rate) #konuşma hızı
        
        self.engine.say(say)
        self.engine.runAndWait()

    
        
    def standBy(self):
        porcupine = pvporcupine.create(
        access_key="8fPjSJyCB9Lb/4UE8Zza3F0/XaC5cE31KbBAjAiEjiTUsbd8UX1OKw==",
        keyword_paths=['jarvis_en_windows_v2_1_0 -.ppn'],
        sensitivities=[0.4])
        
        recoder = PvRecorder(device_index=-1, frame_length=porcupine.frame_length)
        try:
            recoder.start()
            while (True):
                keyword_index = porcupine.process(recoder.read())
                if keyword_index >= 0:
                    recoder.stop()
                    
                    
                    
                    self.authentication()
                    
                    
        except KeyboardInterrupt:
            recoder.stop()
        finally:
            porcupine.delete()
            recoder.delete()



        


        
            


    def authentication(self):
        
        con = sqlite3.connect("user.db")
        cursor = con.cursor()
        cursor.execute("CREATE TABLE IF NOT EXISTS USER(Name TEXT)")
        con.commit()

        
        '''
        self.name = "mert"
            
        i=0

        cursor.execute("insert into USER VALUES(?)",(self.name,))
        con.commit()
        '''
        cursor.execute("select * from USER")
        self.name = cursor.fetchall()
        a=0
        b=0
        while(1):
            self.speak("Hi. Do I know you sir?")
            self.response = sr.Recognizer()

            with sr.Microphone() as source:
                
                
                
                playsound("listeningsound.wav")
                audio = self.response.listen(source,phrase_time_limit=1.5)
            try:
                self.phrase = self.response.recognize_google(audio, language="en-US")
                self.phrase = self.phrase.lower()
                print(self.phrase)
                self.phrase= self.phrase.split(" ")
                if("administration"in self.phrase):
                    self.requests()
                if("yes" in self.phrase):
                    i=0
                    j=0
                    while(i<=1):
                        self.speak("What is your name sir?")
                        
                        with sr.Microphone() as source:
                            
                            playsound("listeningsound.wav")
                            audio = self.response.listen(source,phrase_time_limit=1.5)
                            
                        
                        try:
                            
                            self.phrase = self.response.recognize_google(audio, language="tr-TR")
                            self.phrase = self.phrase.lower()
                            self.phrase =self.phrase.split(" ")
                            print(self.phrase)
                            if("mert" in self.phrase):
                                if(i==1):
                                    self.speak("I remembered you")
                                self.face_recognition()
                            else:
                                if(i==0):
                                    self.speak("This name doesn't sound familiar, can you say it again?")
                                    i+=1
                                    continue
                                if(i==1):
                                    self.speak(f"I don't know you{self.phrase}, i go to standby")
                                    self.standBy()
                                    
                                    
                                
                        except sr.UnknownValueError:
                            
                            if(j==0):
                                self.speak("I cant hear you please talk loudly!")
                                j+=1
                                continue
                            if(j==1):
                                self.speak("You don't answer me, i go to standby")
                                self.standBy()
                            
                        
                        
                if("no"in self.phrase):
                    self.speak("I have to turn standby now ,sorry!")
                    #new member adding
                    self.standBy()
                    

                else:
                    if(a==0):
                        self.speak(" Please only say yes or no!")
                        a+=1
                        continue
                    if(a==1):
                        self.speak("You don't give the answer I want,I go to standBy")
                        self.standBy()
                    

            except sr.UnknownValueError:

                if(b==0):
                    self.speak("I cant hear you please talk loudly!")
                    b+=1
                    continue
                if(b==1):
                    self.speak("You don't answer me, i go to standby")
                    self.standBy()


    def face_recognition(self):



        
        video_capture = cv2.VideoCapture(0)
        img = cv2.imread("dataset/mert.jpg")

        resp = RetinaFace.detect_faces(img, threshold = 0.3)
        
            
        if 1==len(resp):
            for key in resp:
                identity = resp[key]
                facial_area=identity["facial_area"]
                #facial_area_location.append(((facial_area[3],facial_area[0]), (facial_area[1],facial_area[2])))
                
                
                
                crop_img = img[facial_area[1]: facial_area[3]+10, facial_area[0]: facial_area[2]+10]

                

                mert_face_encoding=face_recognition.face_encodings(crop_img)[0]

            # Load a second sample picture and learn how to recognize it.
        
        
            # Create arrays of known face encodings and their names
        known_face_encodings = [  mert_face_encoding ]
        known_face_names = [ "Mert" ]

        face_locations = []
        face_encodings = []
        face_names = []
        process_this_frame = True
        self.speak("I need to confirm you ,please look at the camera.")
        c=0
        d=0
        while c<=50:
            
            # Grab a single frame of video
            ret, frame = video_capture.read()
            # Only process every other frame of video to save time
            if process_this_frame:

                resp = RetinaFace.detect_faces(frame, threshold = 0.3)
                facial_area_location=[]
                
                if 1==len(resp):
                    for key in resp:
                        identity = resp[key]
                        facial_area=identity["facial_area"]
                        facial_area_location.append((facial_area[3],facial_area[0], facial_area[1],facial_area[2]))
                        #3012
                        crop_img = frame[facial_area[1]: facial_area[3], facial_area[0]: facial_area[2]]

                else:
                    crop_img=frame
                
                face_encodings = face_recognition.face_encodings(crop_img)
                face_names = []
                for face_encoding in face_encodings:
                    # See if the face is a match for the known face(s)
                    matches = face_recognition.compare_faces(known_face_encodings, face_encoding,tolerance=0.6)
                    name = "Unknown"
                    # # If a match was found in known_face_encodings, just use the first one.
                    # if True in matches:
                    #     first_match_index = matches.index(True)
                    #     name = known_face_names[first_match_index]
                    # Or instead, use the known face with the smallest distance to the new face
                    face_distances = face_recognition.face_distance(known_face_encodings, face_encoding)
                    best_match_index = np.argmin(face_distances)
                    #if matches[best_match_index]:
                    #    name = known_face_names[best_match_index]
                    #face_names.append(name)

                    if True in matches:
                        first_match_index = matches.index(True)
                        name = known_face_names[first_match_index]
                    face_names.append(name)
            
            c+=1
            process_this_frame = not process_this_frame
            if(c==50):
                self.speak("I can't see your face on camera")
                self.standBy()
            if("Mert" in face_names):
               self.greeting()
            
            if("Unknown" in face_names):
                d+=1

            if(d==6):
                cv2.imwrite("dataset/" +str(datetime)+".jpg",frame)
                self.speak("Unauthorized login procedure, I saved your pictures, don't try to login again")
                self.standBy()
                
            
        


    def greeting(self):
        hour = datetime.now().hour
        if(hour>=7 and hour<12):
            self.speak("Good Morning Sir" )
        elif(hour>=12 and hour<18):
            self.speak("Good Afternoon Sir")
        elif(hour>=18 and hour<22):
            self.speak("Good Evening sir " )
        else:
            self.speak("Good Night sir" )
        
        self.LoggedIn()
        




    def LoggedIn(self):

        self.speak("How are you today?")

        
            
        self.response = sr.Recognizer()

        with sr.Microphone() as source:
                
                
                
            playsound("listeningsound.wav")
            audio = self.response.listen(source,phrase_time_limit=5)
            try:
                self.phrase = self.response.recognize_google(audio, language="en-US")
                self.phrase = self.phrase.lower()
                print(self.phrase)
                self.phrase= self.phrase.split(" ")
                if("good" in self.phrase or "nice" in self.phrase or ("not"and "bad" in self.phrase)):
                    self.happyLogin()
                if("bad"in self.phrase or ("not"and"good" in self.phrase) or "bored" in self.phrase):
                    self.happyLogin()

            except sr.UnknownValueError:
                
                self.sadLogin()



    def happyLogin(self):
        self.speak("What can I do for you sir?")
        self.requests()

    def sadLogin(self):
        self.speak("I think you are bad today,What can I do for you sir?")
        self.requests()
        


    def mainStandBy(self):
        porcupine = pvporcupine.create(
        access_key="8fPjSJyCB9Lb/4UE8Zza3F0/XaC5cE31KbBAjAiEjiTUsbd8UX1OKw==",
        keyword_paths=['jarvis_en_windows_v2_1_0 -.ppn'],
        sensitivities=[0.4])
        
        recoder = PvRecorder(device_index=-1, frame_length=porcupine.frame_length)
        try:
            recoder.start()
            while (True):
                keyword_index = porcupine.process(recoder.read())
                if keyword_index >= 0:
                    recoder.stop()
                    
                    self.happyLogin()
                    
                    
                    
                    
        except KeyboardInterrupt:
            recoder.stop()
        finally:
            porcupine.delete()
            recoder.delete()
        

        

    def requests(self):

        self.response = sr.Recognizer()

        with sr.Microphone() as source:
                
                
                
            playsound("listeningsound.wav")
            audio = self.response.listen(source,timeout=10,phrase_time_limit=8)
            try:
                self.phrase = self.response.recognize_google(audio, language="en-US")
                self.phrase = self.phrase.lower()
                print(self.phrase)
                self.phrase= self.phrase.split(" ")
                if("flip" in self.phrase or "flip" in self.phrase and "money" in self.phrase):
                    self.flip()
                if("joke"in self.phrase):
                    self.joke()
                if("you" in self.phrase and "fast" in self.phrase):
                    self.speak("Sorry for that I decrease the speak rate now")
                    self.rate-=10
                    self.mainStandBy()
                if("you"in self.phrase and"slow" in self.phrase):
                    self.speak("Sorry for that I increase the speak rate now")
                    self.rate+=10
                    self.mainStandBy()
                if("thank" in self.phrase and "you"in self.phrase):
                    self.speak("Your welcome")
                    self.mainStandBy()
                if("I" in self.phrase and "out" in self.phrase and "go"in self.phrase or "exit"in self.phrase):
                    self.speak("See you later ,Sir")
                    self.standBy()
                if("stop" in self.phrase or "play" in self.phrase):
                    self.stopandPlayYoutube()
                if("youtube" in self.phrase):
                    self.speak("What do you want me to open on youtube")
                    self.youtube()
                if("next" in self.phrase):
                    self.nextYoutube()
                if("close" in self.phrase and "youtube" in self.phrase):
                    self.browser.close()
                    self.mainStandBy()
                else:
                    self.speak("I don't know this words")
                    self.mainStandBy()

            except sr.UnknownValueError:
                self.speak("When you want to talk to me you can say Jarvis")
                self.mainStandBy()
            except sr.WaitTimeoutError:
                self.speak("When you want to talk to me you can say Jarvis")
                self.mainStandBy()
                
            



    def flip(self):
        options = ["tura","yazi"]
        answer = random.randint(0,1)
        a=options[answer]
        sentence = f"You got {a}"
        self.speak(sentence)
        self.requests()



    def joke(self):
        self.speak(pyjokes.get_joke())
        self.requests()
        


    def youtube(self):
        self.response = sr.Recognizer()

        with sr.Microphone() as source:
                
                
                
            playsound("listeningsound.wav")
            audio = self.response.listen(source,timeout=10,phrase_time_limit=8)
            try:
                self.phrase = self.response.recognize_google(audio, language="tr-TR")
                self.phrase = self.phrase.lower()
                
                url="https://www.youtube.com/results?search_query={}".format(self.phrase)
                self.options=webdriver.ChromeOptions()
                self.options.add_experimental_option("detach",True)
                self.options.add_experimental_option('excludeSwitches', ['enable-logging'])
                self.browser=webdriver.Chrome(options=self.options)
                
                self.browser.get(url)
                button=self.browser.find_element(By.XPATH,'//*[@id="video-title"]/yt-formatted-string').click()
                self.mainStandBy()
            except sr.UnknownValueError:
                self.mainStandBy()


        
    
    def stopandPlayYoutube(self):
        button=self.browser.find_element(By.XPATH,'//*[@id="movie_player"]/div[1]/video').click()
        self.mainStandBy()
        
    def nextYoutube(self):
        button=self.browser.find_element(By.XPATH,'//*[@class="ytp-next-button ytp-button"]').click()
        self.mainStandBy()

assistant = Voice_Assistant()


